platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/ReportableAsProblem.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.resolution.failure;
+
+import org.gradle.api.problems.internal.InternalProblems;
+
+/**
+ * Represents a {@link Throwable} that can be reported as a {@link org.gradle.api.problems.internal.Problem} to
+ * the {@link org.gradle.api.problems.Problems} service.
+ * <p>
+ * This is necessary, as Dependency Management uses exceptions for flow control, and some exceptions are meant to
+ * be "ignored", in the sense that they don't fail a build and represent dependency failures that can be recovered from.  Due
+ * to this usage, we can't follow the typical advice of creating and reporting a problem at the time we create and
+ * throw the exception.  Instead, exceptions that implement this interface understand how to report themselves as
```

Comment:
🤔 I can see this mechanism being used more widely outside of dependency management. Especially if we plan to migrate this handling higher up outside of dependency-management itself into a build-wide exception handler. I would consider making this text more generic to reflect that. 

Is there an issue to track the work to pull up the exception handling to a higher level? We still miss reporting problems for cases where the user manually throws exceptions gathered from `UnresolvedDependencyResult#getFailure` and `UnresolvedArtifactResult#getFailure`

Reply:
I just created one: https://github.com/gradle/gradle/issues/30547